server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1589,6 +1509,120 @@ static String resolveExpression(String expression, LongSupplier getTime) {
         }
     }
 
+    public static final class ExplicitResourceNameFilter {
+
+        private ExplicitResourceNameFilter() {
+            // Utility class
+        }
+
+        public static List<String> filter(Context context, List<String> expressions) {
+            List<String> result = new ArrayList<>(expressions.size());
+            for (ExpressionList.Expression expression : new ExpressionList(context, expressions)) {
+                validateAliasOrIndex(expression);
+                if (expression.isWildcard() || expression.isExclusion() || ensureAliasOrIndexExists(context, expression.get())) {
+                    result.add(expression.expression());
+                }
+            }
+            return result;
+        }
+
+        @Nullable
+        private static boolean ensureAliasOrIndexExists(Context context, String expression) {
+            boolean ignoreUnavailable = context.getOptions().ignoreUnavailable();
+            IndexAbstraction indexAbstraction = context.getState().getMetadata().getIndicesLookup().get(expression);
+            if (indexAbstraction == null) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    throw notFoundException(expression);
```

Comment:
Remind me again ... if we have "foo*,-foo1" and foobar exists in the cluster but foo1 does not exist and ignoreUnavailable is false should that throw the not found exception ?

Reply:
No, we shouldn't throw the not found exception because exclusions, i.e. `-foo1`, never do it.
Conversely, `foo*,foo1` will throw the exception. Also, confusing, `-foo1,foo*` should throw the "not found" exception (for `-foo1`).